pycobertura/filesystem.py

```diff
@@ -68,3 +71,52 @@ def open(self, filename):
             f.close()
         except KeyError:
             raise self.FileNotFound(filename)
+
+
+class GitFileSystem(FileSystem):
+    def __init__(self, repo_folder, ref):
+        self.repository = repo_folder
+        self.ref = ref
+        self.repository_root = self._get_root_path(repo_folder)
+        self.prefix = self.repository.replace(self.repository_root, "").lstrip("/")
+
+    def real_filename(self, filename):
+        prefix = "{}/".format(self.prefix) if self.prefix else ""
+        return "{ref}:{prefix}{filename}".format(
+            prefix=prefix, ref=self.ref, filename=filename
+        )
+
+    def has_file(self, filename):
+        command = "git --no-pager show {}".format(self.real_filename(filename))
+        return_code = subprocess.call(
+            command,
+            cwd=self.repository,
+            shell=True,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+        return not bool(return_code)
+
+    def _get_root_path(self, repository_folder):
+        command = "git rev-parse --show-toplevel"
+        output = subprocess.check_output(shlex.split(command), cwd=repository_folder)
+        return output.decode("utf-8").rstrip()
+
+    @contextmanager
+    def open(self, filename):
+        """
+        Yield a file-like object for file `filename`.
+
+        This function is a context manager.
+        """
+        filename = self.real_filename(filename)
+
+        command = "git --no-pager show {}".format(filename)
+
+        try:
+            output = subprocess.check_output(shlex.split(command), cwd=self.repository)
+        except (OSError, subprocess.CalledProcessError):
```

Comment:
Do we still need `OSError` here now that we use `subprocess`? In what circumstances would we expect `OSError` to get raised?

Reply:
Well, according to https://docs.python.org/2/library/subprocess.html#exceptions
we should catch OSError (on systems where git is not available) and CalledProcessError (if the executable returns a non-zero error code).
I think this is a good idea to keep both.

Reply:
Oh great, thanks for looking into it!